### PR TITLE
Editing Roads

### PR DESCRIPTION
**Pull request by vicr123**
This PR adds support for editing a road in Select mode

### CODE REVIEW

File: BnbnavNetClient/Services/EditControllers/SelectEditController.cs [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+using BnbnavNetClient.Models;
+using BnbnavNetClient.ViewModels;
 using BnbnavNetClient.Views;
 
 namespace BnbnavNetClient.Services.EditControllers;
 
 public class SelectEditController : EditController
 {
+    private readonly MapEditorService _mapEditorService;
+    private Point _initialPointerPosition;
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/EditControllers/SelectEditController.cs]: no explicit default modifiers

File: BnbnavNetClient/Services/EditControllers/SelectEditController.cs [hunk 2]
```diff
@@ -23,6 +33,29 @@ public override void PointerMoved(MapView mapView, PointerEventArgs args)
 
     public override void PointerReleased(MapView mapView, PointerReleasedEventArgs args)
     {
+        var pointerPos = args.GetPosition(mapView);
+
+        if (new ExtendedLine
+            {
+                Point1 = pointerPos,
+                Point2 = _initialPointerPosition
+            }.Length >= 0.5) return;
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/EditControllers/SelectEditController.cs]: consider making a constructor which takes these as arguments; using a distance function straight away or even writing it manually is probably clearer. if you are to continue with this method, please format it more sensibly.
```csharp
if (new ExtendedLine { Point1 = pointerPos, Point2 = _initialPointerPosition }.Length >= 0.5)
    return;
```
or declare and define the line in a variable before

File: BnbnavNetClient/Services/EditControllers/SelectEditController.cs [hunk 3]
```diff
@@ -23,6 +33,29 @@ public override void PointerMoved(MapView mapView, PointerEventArgs args)
...
+        {
+            case Edge edge:
+                mapView.OpenFlyout(new RoadEditViewModel(_mapEditorService, edge.Road));
+                break;
+            case Node node:
+            {
+                if (_mapEditorService.MapService!.Landmarks.Values.Any(x => x.Node == node))
+                {
+                    mapView.OpenFlyout(new LandmarkFlyoutViewModel(_mapEditorService, node));
+                }
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/EditControllers/SelectEditController.cs]: use no braces, consistently with rest of projcet

File: BnbnavNetClient/Services/EditControllers/SelectEditController.cs [hunk 4]
```diff
@@ -23,6 +33,29 @@ public override void PointerMoved(MapView mapView, PointerEventArgs args)
...
+            }.Length >= 0.5) return;
+            
+        var item = mapView.HitTest(pointerPos).FirstOrDefault(x => x is Edge or Node);
+        switch (item)
+        {
+            case Edge edge:
+                mapView.OpenFlyout(new RoadEditViewModel(_mapEditorService, edge.Road));
+                break;
+            case Node node:
+            {
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/EditControllers/SelectEditController.cs]: no braces on cases

File: BnbnavNetClient/Services/NetworkOperations/NetworkOperation.cs [hunk 5]
```diff
@@ -12,7 +12,7 @@ public abstract class NetworkOperation
     public abstract Task PerformOperation();
     public abstract void Render(MapView mapView, DrawingContext context);
     
-    public IList<MapItem> ItemsNotToRender { get; set; } = new List<MapItem>();
+    public List<MapItem> ItemsNotToRender { get; set; } = new List<MapItem>();
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/NetworkOperations/NetworkOperation.cs]: target-type new()

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        ItemsNotToRender.AddRange(_affectedEdges);
+        
+        try
+        {
+            (await _editorService.MapService!.Submit($"/roads/{_road.Id}", new
+            {
+                name = _newName,
+                type = _newRoadType.ServerName()
+            })).AssertSuccess();
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:39]: consider setting to a variable

File: BnbnavNetClient/Views/MapView.axaml.cs [hunk 7]
```diff
@@ -92,18 +93,21 @@ protected override void OnInitialized()
...
+                            new()
                             {
-                                MapViewModel.QueueDelete(node);
-                            })
+                                Header = _i18n["NODE_DELETE"],
+                                Command = ReactiveCommand.Create(() =>
+                                {
+                                    MapViewModel.QueueDelete(node);
+                                })
+                            }
```
Comment (Vrabbers) [inline BnbnavNetClient/Views/MapView.axaml.cs:110]: consider using some converter/data template here instead, though this works

**Review by Vrabbers — CHANGES_REQUESTED** [review]

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Avalonia.Media;
+using BnbnavNetClient.Models;
+using BnbnavNetClient.Views;
+
+namespace BnbnavNetClient.Services.NetworkOperations;
+
```
Comment (Vrabbers) [inline BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:10]: remove this newline

Comment (Vrabbers) [inline BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:10]: remove this newline

File: BnbnavNetClient/ViewModels/RoadEditViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+using Avalonia.Collections;
+using Avalonia.Controls.Primitives;
+using BnbnavNetClient.Models;
+using BnbnavNetClient.Services;
+using BnbnavNetClient.Services.NetworkOperations;
+using ReactiveUI;
+using ReactiveUI.Fody.Helpers;
+
+namespace BnbnavNetClient.ViewModels;
+
```
Comment (Vrabbers) [inline BnbnavNetClient/ViewModels/RoadEditViewModel.cs:13]: no newline

File: BnbnavNetClient/Views/RoadEditView.axaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+                    <TextBlock Text="{Binding HumanReadableName}" />
+                </DataTemplate>
+            </ComboBox.ItemTemplate>
+        </ComboBox>
+        <StackPanel Margin="0 4 0 0" HorizontalAlignment="Right" Orientation="Horizontal">
+            <Button Command="{Binding CancelClicked}" Content="{str:Tr Key=CANCEL}" />
+            <Button Classes="accent" IsEnabled="{Binding UpdateButtonEnabled}" Command="{Binding UpdateClicked}" Content="{str:Tr Key=UPDATE}" />
+        </StackPanel>
+    </StackPanel>
+</UserControl>
```
Comment (Vrabbers) [inline BnbnavNetClient/Views/RoadEditView.axaml:27]: format with 2 spaces rather than 4

File: BnbnavNetClient/Views/RoadEditView.axaml.cs [hunk 11]
```diff
@@ -0,0 +1,18 @@
+using Avalonia;
+using Avalonia.Controls;
+using Avalonia.Markup.Xaml;
+
+namespace BnbnavNetClient.Views;
+
```
Comment (Vrabbers) [inline BnbnavNetClient/Views/RoadEditView.axaml.cs:6]: newline

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs (hunk 6 above)
Comment (vicr123) [reply BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:39]: There's nothing to set to a variable here apart from the response which I think should be inline if we are only ensuring that it is successful

File: BnbnavNetClient/Views/MapView.axaml.cs (hunk 7 above)
Comment (vicr123) [reply BnbnavNetClient/Views/MapView.axaml.cs:110]: This can be refactored later by someone who knows what they are doing

**Review by Martysh12 — APPROVED** [review]
Well errr, I suppose it looks good to me?

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        try
+        {
+            (await _editorService.MapService!.Submit($"/roads/{_road.Id}", new
+            {
+                name = _newName,
+                type = _newRoadType.ServerName()
+            })).AssertSuccess();
+        }
+        catch (HttpRequestException)
+        {
```
Comment (Royce551) [inline BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:42]: i'm guessing handling for this is planned?

File: BnbnavNetClient/Services/NetworkOperations/NodeMoveOperation.cs [hunk 13]
```diff
@@ -53,11 +53,7 @@ public override void Render(MapView mapView, DrawingContext context)
         var movingRect = _toUpdate.BoundingRect(mapView);
         var movedRect = _updateTo.BoundingRect(mapView);
 
-        ExtendedLine lineBetween = new()
-        {
-            Point1 = movingRect.Center,
-            Point2 = movedRect.Center
-        };
+        ExtendedLine lineBetween = new(movingRect.Center, movedRect.Center);
```
Comment (Royce551) [inline BnbnavNetClient/Services/NetworkOperations/NodeMoveOperation.cs]: ask vrabb about this, but IMO it's better to var lineBetween = new ExtendedLine(. . .)

**Review by Royce551 — APPROVED** [review]
lgtm

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs (hunk 12 above)
Comment (vicr123) [reply BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:42]: There's nothing to handle right now but if we want to show an error message to the user that's where we'll raise it

File: BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs (hunk 6 above)
Comment (vicr123) [reply BnbnavNetClient/Services/NetworkOperations/RoadUpdateOperation.cs:39]: Resolving now - if we want to action this change we should apply it to all operations simultaneously in another PR

**Pull request merged**